src/core/Akka/Dispatch/CurrentSynchronizationContextDispatcher.cs

```diff
@@ -0,0 +1,98 @@
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using Akka.Actor;
+using Akka.Configuration;
+
+namespace Akka.Dispatch
+{
+    /// <summary>
+    /// INTERNAL API
+    /// 
+    /// Produces <see cref="ExecutorService"/> that dispatches messages on the current synchronization context,
+    ///  e.g. WinForms or WPF GUI thread
+    /// </summary>
+    internal sealed class CurrentSynchronizationContextExecutorServiceFactory : ExecutorServiceConfigurator
+    {
+        public override ExecutorService Produce(string id)
+        {
+            return new TaskSchedulerExecutor(id, TaskScheduler.FromCurrentSynchronizationContext());
+        }
+
+        public CurrentSynchronizationContextExecutorServiceFactory(Config config, IDispatcherPrerequisites prerequisites) : base(config, prerequisites)
+        {
+        }
+    }
+
+    /// <summary>
+    /// Used to create instances of the <see cref="PinnedDispatcher"/>. 
+    /// 
+    /// Each actor created using the pinned dispatcher gets its own unique thread.
+    /// <remarks>
+    /// Always returns a new instance.
+    /// </remarks>
+    /// </summary>
+    internal sealed class CurrentSynchronizationContextDispatcherConfigurator : MessageDispatcherConfigurator
+    {
+        private readonly ExecutorServiceConfigurator _executorServiceConfigurator;
+
+        public CurrentSynchronizationContextDispatcherConfigurator(Config config, IDispatcherPrerequisites prerequisites)
+            : base(config, prerequisites)
+        {
+
+            _executorServiceConfigurator = new CurrentSynchronizationContextExecutorServiceFactory(config, prerequisites);
+            // We don't bother trying to support any other type of exectuor here. PinnedDispatcher doesn't support them
+        }
+
+        public override MessageDispatcher Dispatcher()
+        {
+            return new CurrentSynchronizationContextDispatcher(this, Config.GetString("id"),
+                Config.GetInt("throughput"),
+                Config.GetTimeSpan("throughput-deadline-time").Ticks,
+                _executorServiceConfigurator,
+                Config.GetTimeSpan("shutdown-timeout"));
+        }
+    }
+
+    /// <summary>
+    /// Behaves like a <see cref="PinnedDispatcher"/> and always executes using <see cref="CurrentSynchronizationContextExecutorServiceFactory"/>
+    /// </summary>
+    public sealed class CurrentSynchronizationContextDispatcher : Dispatcher
+    {
+        public CurrentSynchronizationContextDispatcher(MessageDispatcherConfigurator configurator, string id, int throughput, long? throughputDeadlineTime, ExecutorServiceFactory executorServiceFactory, TimeSpan shutdownTimeout) 
+            : base(configurator, id, throughput, throughputDeadlineTime, executorServiceFactory, shutdownTimeout)
+        {
+            /*
+             * Critical: in order for the CurrentSynchronizationContextExecutor to function properly, it can't be lazily 
```

Comment:
This is where we really fixed the bug; normally the `ExecutorService` is lazily initiated. This won't work - we need to activate the `ExecutorService` immediately so the correct `SynchronizationContext` is captured. So we do that by just running a no-op task at the tail end of the constructor. This is safe because all of the code needed for this method to run is constructed in the base class constructor, so there's no possibility of a race condition there.